File: backend/src/main/kotlin/fourkeymetrics/metrics/model/Metrics.kt
package fourkeymetrics.metrics.model

import com.fasterxml.jackson.databind.annotation.JsonSerialize
import fourkeymetrics.common.serializer.NumberSerializer


enum class LEVEL {
    ELITE, HIGH, MEDIUM, LOW, INVALID
}

data class Metrics(
    @JsonSerialize(using = NumberSerializer::class)
    val value: Number,
    val level: LEVEL?,
    val startTimestamp: Long,
    val endTimestamp: Long
) {
    constructor(value: Number, startTimestamp: Long, endTimestamp: Long) : this(
        value,
        null,
        startTimestamp,
        endTimestamp
    )
}

File: frontend/src/models/metrics.ts
export enum MetricsType {
	DEPLOYMENT_FREQUENCY = "Deployment Frequency",
	LEAD_TIME_FOR_CHANGE = "Lead Time for Change",
	MEAN_TIME_TO_RESTORE = "Mean Time To Restore",
	CHANGE_FAILURE_RATE = "Change Failure Rate",
}

export enum MetricsLevel {
	ELITE = "ELITE",
	HIGH = "HIGH",
	INVALID = "INVALID",
	LOW = "LOW",
	MEDIUM = "MEDIUM",
}

export enum MetricsUnit {
	FORTNIGHTLY = "Fortnightly",
	MONTHLY = "Monthly",
}

export type MetricType = "df" | "lt" | "mttr" | "cfr";

export interface Metrics {
	value: number | "NaN" | undefined;
	endTimestamp: number;
	startTimestamp: number;
}

export interface MetricsSummary extends Metrics {
	level: MetricsLevel;
}

export interface ValidMetric extends Metrics {
	value: number;
}

export interface MetricsInfo {
	details: Metrics[];
	summary: MetricsSummary;
}